Userland/Libraries/LibGUI/TextDocument.cpp

```diff
@@ -384,6 +384,99 @@ DeprecatedString TextDocument::text_in_range(TextRange const& a_range) const
     return builder.to_deprecated_string();
 }
 
+// This function will return the position of the previous grapheme cluster
+// break, relative to the cursor, for "correct looking" parsing of unicode based
+// on grapheme cluster boundary algorithm.
+size_t TextDocument::get_previous_grapheme_cluster_boundary(
+    TextPosition const& cursor) const
```

Comment:
Some of the formatting is still quite weird. For example, this should be one line. Same comment on many places in these changes.